### DISCUSSION TITLE
Wails and Bootstrap

### DISCUSSION THREAD

**Discussion by ahanniga** [wailsapp/wails#2636] [Q&A] [answered]
I'm using Bootstrap 5 in my Wails/Svelte app and all works well in dev mode. 

However when built I get javascript errors because the global variable `bootstrap` is undefined. The UI looks fine. I've tried playing with the `//go:embed` in `main.go` without success.

Can anyone suggest config that will make `bootstrap` var available outside of `wails dev` ?

Thanks!

Aidan

**Comment by misitebao** [comment]
This problem has nothing to do with Wails, it should be your front-end code problem without accident. Can you provide your code and error message?

**Reply by ahanniga** [reply]
The dependencies are in `main.js` as follows:

```js
import './bootstrap.bundle'
import './bootstrap.css'
import './assets/bootstrap-icons/bootstrap-icons.css'
import './style.css'
import App from './App.svelte'

const app = new App({
  target: document.getElementById('app')
})

export default app
```

These load fine and everything works except the build executable cannot access the bootstrap environment (ignore the TLS error):

![image](https://user-images.githubusercontent.com/28892939/235361123-7d6e0ee9-3cba-4f64-817f-3f05cfedd7a5.png)

In `wails dev` mode:

![image](https://user-images.githubusercontent.com/28892939/235361262-8b422228-85f4-47a3-9a7a-25fcb3acae51.png)

Admittedly, this only limits my ability to use commands such as `bootstrap.Modal.getOrCreateInstance()` but it would be great to resolve if possible.

**Reply by misitebao** [reply]
The reason why global variables can be accessed in dev but not in build is because build obfuscates and tree shakes the code.

**Comment by KiddoV** [comment]
Maybe try to "inject" bootstrap like I did with jQuery here: https://github.com/wailsapp/wails/discussions/1970#discussioncomment-3923734

**Reply by ahanniga** [reply]
Looks promising - thanks. Will try that.

**Comment by ahanniga** [comment]
The only way I've found that works is linking the bootstrap CDN in index.html. Shouldn't the resources imported in `main.js` be copied into dist/ by the build?

**Reply by misitebao** [reply]
It is not recommended to use CDN resources in the UI interface of wails, but to package all the resources used. Because as an independent APP, the UI should be able to display normally even when there is no network, instead of crashing or failing to display the interface due to lack of resources.

**Reply by ahanniga** [reply]
Absolutely. I only use the `bootstrap` variable to launch dialogs. If you don't consider it a bug, I'll find a work-around, thanks.

**Comment by ahanniga** [comment, marked as answer]
Here is a ~~hack~~ workaround that does not require the `bootstrap` global variable.

1. Place the Bootstrap files in `frontend/src/assets`
2. In `main.js`, import them:

```javascript
import './assets/bootstrap.bundle.min'
import './assets/bootstrap.min.css'
import './assets/style.css'
import App from './App.svelte'
```

3. Create a Svelte component such as `Dialogs.svelte` with some hidden links. When `EventsEmit("evLoginDialog")` is called the dialog is invoked via the `data-bs-toggle` and `data-bs-target` attribtes.

```svelte
<script>
    import Login from "./Login.svelte";

    const onLoginDialog = () => {
        document.getElementById('launchLoginDialog').click();
    }
    window.runtime.EventsOn('evLoginDialog', onLoginDialog);

</script>

<a id="launchLoginDialog" data-bs-toggle="modal" data-bs-target="#loginDialog"></a>

<Login />
```
